File: plugins/collector/event/apps/bigdata/flink.go
package bigdata

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

// Totally work on argv, unreliable maybe, but useful
type Flink struct {
	version    string
	reg        *regexp.Regexp
	versionReg *regexp.Regexp
}

func (Flink) Name() string { return "flink" }

func (Flink) Type() string { return "bigdata" }

func (f Flink) Version() string { return f.version }

func (f *Flink) Match(p *process.Process) bool {
	if p.Name != "java" {
		return false
	}
	if f.reg == nil {
		f.reg = regexp.MustCompile(` org\.apache\.flink\.[a-zA-Z\.]+`)
	}
	if s := f.reg.FindString(p.Argv); s == "" {
		return false
	}
	return true
}

func (f *Flink) Run(p *process.Process) (m map[string]string, err error) {
	// temporary files like /tmp/hsperfdata_(username)
	if f.versionReg == nil {
		f.versionReg = regexp.MustCompile(`flink-dist-(\d+\.)+\d\.jar`)
	}
	if s := f.versionReg.FindString(p.Argv); s != "" {
		f.version = strings.TrimRight(strings.TrimPrefix(s, "flink-dist-"), ".jar")
	}
	// jobmanager or taskmanager
	m = make(map[string]string)
	if strings.Contains(p.Argv, "jobmanager.") {
		m["run_mode"] = "jobmanager"
	} else if strings.Contains(p.Argv, "taskmanager.") {
		m["run_mode"] = "taskmanager"
	} else {
		m["run_mode"] = "unknown"
	}

	return
}

func init() { apps.Regist(&Flink{}) }

File: plugins/collector/src/event/apps/bigdata/flink.rs
use crate::event::apps::{find_version, jar_names_for_pid, AppProc, IApp};
use std::collections::HashMap;

pub struct Flink {
    version: String,
}

impl Flink {
    pub fn new() -> Self { Flink { version: String::new() } }
}

impl IApp for Flink {
    fn name(&self)     -> &'static str { "flink" }
    fn app_type(&self) -> &'static str { "bigdata" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "java" && p.argv.contains("org.apache.flink.")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        self.version.clear();
        // "flink-dist-1.15.2.jar"
        for jar in jar_names_for_pid(p.pid) {
            if jar.starts_with("flink-dist-") && jar.ends_with(".jar") {
                let mid = jar.trim_start_matches("flink-dist-").trim_end_matches(".jar");
                let ver = find_version(mid);
                if !ver.is_empty() { self.version = ver; break; }
            }
        }
        let mut m = HashMap::new();
        let run_mode = if p.argv.contains("jobmanager.") {
            "jobmanager"
        } else if p.argv.contains("taskmanager.") {
            "taskmanager"
        } else {
            "unknown"
        };
        m.insert("run_mode".into(), run_mode.into());
        Ok(m)
    }
}
